### DISCUSSION TITLE
How to build a SSR cache layer?

### DISCUSSION THREAD

**Discussion by thijsw** [nuxt/nuxt#16423] [Questions]
I am using Nuxt 3 with SSR enabled and it takes quite some time to build the page server-side due to various API calls that need to be performed. I'd like to make a cache layer in Nuxt or Nitro that inspects IP address, Cookies, User Agent and other variables and can decide whether to serve a cached copy from Redis or generate a fresh SSR response. Ideally, I'd like to use the built-in Nuxt server and no other reverse proxies like nginx, so that I can keep the caching logic in Javascript.

In the Nitro documentation I see that there is a [Cache API](https://nitro.unjs.io/guide/introduction/cache) and a [Storage Layer](https://nitro.unjs.io/guide/introduction/storage), but I am not aware of how to leverage these API in Nuxt? Maybe there is a hook I can use?

**Comment by Diizzayy** [comment]
@thijsw You're on the right track, The [Nitro Storage Layer](https://nitro.unjs.io/guide/introduction/storage) can be implemented by configuring the redis driver in your nuxt config, then defining your required logic in the api routes to discern when data should be retrieved from / stored to redis. Example below:

`nuxt.config.ts`
```ts
export default defineNuxtConfig({
  nitro: {
    storage: {
      'redis': {
        driver: 'redis',
        username: "",
        password: "",
        host: "127.0.0.1",
        port: 6379,
      }
    }
  }
})
```


`/nuxt-app/server/api/demo.ts`
```ts
export default defineEventHandler(async (event) => {
  // retrieve user agent
  const ua = event?.req?.headers?.['user-agent']

  // retrieve a user cookie
  let userData = getCookie(event, '<cookie_name>')

  // your custom logic to conditionally retrieve data
  // based on cookie, ip address, etc...

  // set redis cache
  await useStorage().setItem('redis:demo', { hello: 'world' })

  // update redis cache
  const data = await useStorage().getItem('redis:demo')

  return data
})
```

**Reply by thijsw** [reply]
Thank you for your answer! This is very helpful and not documented as far as I can see. As for my case, I am more interested in caching the whole SSR response, not just individual API calls, as most of them are outside my control (I could proxy them ofcourse)

**Reply by NicoFerna** [reply]
Hi @Diizzayy, thanks for sharing your example. Could you explain how to move this caching logic in a reusable component. I tried with composables but useStorage is not available, is this a case for a nitro plugin?

**Comment by bhaskarGyan** [comment]
@thijsw , If you want to cache the whole SSR rendered response and then serve it based on the headers condition and/or device type(mobile, desktop, crawler, etc) you can use the below package.

https://www.npmjs.com/package/nuxt-cache-ssr

Currently, it will store the page **InMemory**,
storing in **Redis** is **in the pipeline.**

you can set the cache key based on any of the headers 


 ```
 key:(route:string,headers: any,device:Device)=>{  
     
     //sample of using device type to generate key
      const {userAgent,...deviceType} = device
      const key = [route];
      Object.keys(deviceType).forEach(val => {
        if(deviceType[val]){
          key.push(val)
        }
      })
       // returned value will be hashed using ohash
      return key.join("-")
    }
```

or you can also use out-of-the-box device detection based on [nuxt-community/device-module](https://github.com/nuxt-community/device-module/blob/master/lib/plugin.js)

```
interface Device {
  userAgent: string
  isDesktop: boolean
  isIos: boolean
  isAndroid: boolean
  isMobile: boolean
  isMobileOrTablet: boolean
  isDesktopOrTablet: boolean
  isTablet: boolean
  isWindows: boolean
  isMacOS: boolean
  isApple: boolean
  isSafari: boolean
  isFirefox: boolean
  isEdge: boolean
  isChrome: boolean
  isSamsung: boolean
  isCrawler: boolean
}
```

**Reply by thijsw** [reply]
Thank you for your reply and what a coincidence that you were building this module the past days, will definitely check this out!

**Reply by bhaskarGyan** [reply]
I was facing similar issue in my app and there was no package similar to [nuxt-perfect-cache](https://www.npmjs.com/package/nuxt-perfect-cache) or [nuxt-ssr-cache](https://www.npmjs.com/package/nuxt-ssr-cache) for Nuxt3. 
So thought to give it a shot.

[Hybrid Rendering (Route Caching rules)](https://github.com/nuxt/framework/discussions/560) is also an option but as of now, it won't give cache control based on device type/ headers.


In case you are facing any issues/ missing feature requests, please create an issue in the GitHub repo.

**Comment by Baroshem** [comment]
https://twitter.com/jacobandrewsky/status/1590642026458255360?s=20&t=S6tfAxyguxkiS3C4rFBu9A

**Reply by Baroshem** [reply]
Let me know if such interface would work well for you and I can start working on a Proof of Concept solution :)

**Reply by thijsw** [reply]
Ideally, I'd like to have some callback methods so you can precisely determine which pages should be cached and under which key (so you could disinguish output for crawlers and normal users for example), something like: 

```typescript
export default defineNuxtConfig({
  // modules: ....

  serverCache: {
    // Determine whether we should cache the output of this route or not
    cache(route) {
      return true
    },

    // Have complete control of the cache key 
    key(route) {
      return route.path
    }
  }
})
```

**Reply by Baroshem** [reply]
If I understand correctly, you would like to have them registered as a part of the module configuration?

Like:
```ts
export default defineNuxtConfig({
  // modules: ....

  serverCache: {
     routes: [
       {
           path: '/category',
           key: 'category',
           handler: () => {}
       }
    ]
  }
})
```

Is it correct?

**Reply by thijsw** [reply]
Well, yeah, that's another way to put it. But I guess most of the time you have some pages that you would want to cache, and some not, defining the keys per route might be too verbose in most cases.

For example, in the case of a news blog, you might have URLs in the form `/article/:id` you want to cache and other pages, such as `/admin` you wouldn't cache. Similarly, search engines are allowed to index the whole articles while anonymous users may only read a part of them. I thought a syntax like this could make this work:

```typescript
export default defineNuxtConfig({
  // modules: ....

  serverCache: {
    // Determine whether we should cache the output of this route or not
    cache(route) {
      if (route.path.startsWith('/article/') {
        return true
      }

      // more rules ...

      return false
    },

    // Have complete control of the cache key 
    key(route) {
      // determine whether user is a crawler
     const prefix = crawler ? 'crawler:' : 'public:'

      return prefix + route.path
    }
  }
})
```

**Reply by Baroshem** [reply]
Yeah, I completely get that. In my case, I wanted to define useServerCache in certain pages rather than in global config but we can support both as well. By following this approach, you can decide whether to cache certain page or not. 

But by supporting both scenarios, you can register cache roules in configuration and inside certain pages.

By looking at your syntax, I think that you could benefit from @bhaskarGyan module.

The idea I had in mind was very basic one so that the users would not have that much things to configure in order to make it work.

**Reply by thijsw** [reply]
Oh great, I was not aware that you could leverage a composable to control the caching of a page. It makes a lot of sense of course, because you can distinguish between routes more naturally. I am looking forward to your implementation!

**Reply by Baroshem** [reply]
I will try to build something and I will see how it works. If my solution will make any sense then we could join forces with @bhaskarGyan and try to merge this into one module under nuxt-modules organization :)

**Reply by bhaskarGyan** [reply]
Sure @Baroshem , Looking forward to your solution.

Meanwhile, I am also looking into adding the redis cache by leveraging unstorage this week

**Reply by bhaskarGyan** [reply]
@thijsw , you can `return false` from the custom key function (implemented in [v1.1.0](https://www.npmjs.com/package/nuxt-cache-ssr/v/1.1.0)) to skip cache for any particular route / subroute defined in `pages` array.
The custom key function already has the device type info along with headers, so you can customise the key for crawlers/ web etc based on the need
